aten/src/ATen/native/mkldnn/Conv.cpp

```diff
@@ -35,43 +35,92 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
-    IntArrayRef input_size, IntArrayRef weight_size,
-    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
-{
-  auto dim = input_size.size();
-  std::vector<int64_t> output_size(dim);
-  output_size[0] = input_size[input_batch_size_dim];
-  output_size[1] = weight_size[weight_output_channels_dim];
-  for (size_t d = 2; d < dim; ++d) {
-    auto kernel = dilation[d - 2] * (weight_size[d] - 1) + 1;
-    output_size[d] = (input_size[d] + (2 * padding[d - 2])
-                        - kernel) / stride[d - 2] + 1;
+  const ideep::tensor& x = itensor_from_mkldnn(input);
+  const ideep::tensor& w = itensor_from_mkldnn(weight);
+
+
+  std::vector<int64_t> kernel_size(input.dim());
+  if (groups > 1 && (weight.dim() == input.dim() + 1)) {
+    kernel_size[0] = weight.size(0) * weight.size(1);
+    std::copy_n(
+        weight.sizes().cbegin() + 2, input.dim() - 1, kernel_size.begin() + 1);
+  } else {
+    std::copy_n(
+        weight.sizes().cbegin(), input.dim(), kernel_size.begin());
+  }
+
+  std::vector<int64_t> output_sizes = conv_output_size(
+    input.sizes(), kernel_size, padding, stride, dilation);
+
+  ideep::scale_t dummy_scale;
+  ideep::tensor y;
+  if (bias.defined()) {
+    const ideep::tensor& b = itensor_from_mkldnn(bias);
+    ideep::convolution_forward::compute<AllocForMKLDNN>(
+      x,
+      w,
+      b,
+      {output_sizes.cbegin(), output_sizes.cend()},
+      y,
+      {stride_vec.cbegin(), stride_vec.cend()},
+      {dilation_vec.cbegin(), dilation_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      groups,
+      dummy_scale,
+      dummy_scale,
+      dummy_scale,
+      ideep::descriptor_group::attr_t{},
+      ideep::algorithm::convolution_direct,
+      ideep::prop_kind::forward);
+  } else {
+    ideep::convolution_forward::compute<AllocForMKLDNN>(
+      x,
+      w,
+      {output_sizes.cbegin(), output_sizes.cend()},
+      y,
+      {stride_vec.cbegin(), stride_vec.cend()},
+      {dilation_vec.cbegin(), dilation_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      groups,
+      dummy_scale,
+      dummy_scale,
+      dummy_scale,
+      ideep::descriptor_group::attr_t{},
+      ideep::algorithm::convolution_direct,
+      ideep::prop_kind::forward);
   }
-  return output_size;
+  return new_with_itensor_mkldnn(std::move(y), input.options());
 }
 
 at::Tensor mkldnn_convolution(
     const at::Tensor& input, const at::Tensor& weight, const at::Tensor& bias,
     IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
 {
-  auto output = at::empty(conv_output_size(
-    input.sizes(), weight.sizes(), padding, stride, dilation, groups), input.options());
+  if (input.type_id() == MkldnnCPUTensorId()) {
+    return _ideep_conv2d(input, weight, bias, padding, stride, dilation, groups);
```

Comment:
can you explain the difference between these two (besides the tensor type)?  Is it because we call ideep here and mkldnn below?  Should we just use one or the other?

Reply:
Yes I agree we should use one notation, will rename it to `_mkldnn_conv2d`